core/src/main/scala/kafka/server/KafkaServer.scala

```diff
@@ -249,6 +247,10 @@ class KafkaServer(val config: KafkaConfig, time: Time = SystemTime, threadNamePr
         registerStats()
 
         brokerState.newState(RunningAsBroker)
+
+        /* start kafka controller after setting state RunningAsBroker as it subsequently become controller */
+        kafkaController.startup()
```

Comment:
@ijuma, from my perspective, RunningAsController is not a useful state and instead should be separate boolean attribute as you said.  To the extent that clients are using this state to know whether a broker is "running", they don't care if it's the controller or not.  I just don't know what the consequences would be of removing that state.

@junrao?